First Submission

malber71 opened the pull request:
My first submission

malber71 committed:
add Api call and getInitialCards

malber71 committed:
created edit avatar modal, connected other apis for posting profile information, adding cards, stopped at avatar edit modal, need to work on modal functionality and close btn

malber71 committed:
created delete modal with formatting, started code to handle buttons

malber71 committed:
enabled delete function, getting errors that need to be corrected

malber71 committed:
finished most of project, need to correct mobile edit avatar button

malber71 committed:
ready for first submission

malber71 committed:
ready for first submission

malber71 committed:
ready for first submission

sebasLTripleTen commented on src/utils/Api.js:
```diff
@@ -0,0 +1,106 @@
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then((res) => {
```
To **avoid repeating response-checking logic** across multiple API requests, you should define a reusable method called `_checkResponse`. This method should handle checking whether the server response is successful (`res.ok`) and process or reject it accordingly:

```js
_checkResponse(res) {
    if (res.ok) {
        return res.json();
    }
    return Promise.reject(`Error: ${res.status}`);
}
```

Once defined, you can simplify your API calls by using:

```js
.then(this._checkResponse)
```

Note: Make sure to pass a reference to the method (i.e., without parentheses), not the result of calling it. This ensures that the function is only executed when the .then() block is triggered.

malber71 replied:
fixed duplication

sebasLTripleTen commented on src/utils/Api.js:
```diff
@@ -0,0 +1,106 @@
+    });
+  }
+  getUserInfo() {
+    return fetch(`${this._baseUrl}/users/me`, {
```
**COULD BE IMPROVED**

To avoid repeating the same fetch-and-check pattern in every request, consider creating a dedicated method that handles both fetching and response checking:

```js
_request(url, options) {
  return fetch(url, options).then(this._checkResponse);
}
```

You can then replace direct fetch calls with `this._request(...)` throughout your class. This keeps your code cleaner and easier to maintain.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -0,0 +1,446 @@
+    })
+    .catch(console.error)
+    .finally(() => {
+      closeModal(editProfileModal);
```
The modal should only be closed after a successful response from the server. This means the call to `closeModal()` should be placed inside the `.then()` block of your promise chain.

Avoid placing it inside a `.finally()` block, doing so would close the modal even if the request fails, which forces the user to reopen it and re-enter the data. Keeping the modal open after an error allows users to quickly fix their input and retry without losing context.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -0,0 +1,446 @@
+    })
+    .catch(console.error)
+    .finally(() => {
+      closeModal(editProfileModal);
```
You are currently closing the Edit Profile modal in this block, but the associated request does not update the profile information.

Please remove this line.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -0,0 +1,446 @@
+    })
+    .catch(console.error)
+    .finally(() => {
+      closeModal(editProfileModal);
```
You are currently closing the Edit Profile modal in this block, but the associated request does not update the profile information.

Please remove this line.

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -0,0 +1,446 @@
+    })
+    .catch(console.error)
+    .finally(() => {
+      closeModal(editProfileModal);
```
You are currently closing the Edit Profile modal in this block, but the associated request does not update the profile information.

Please remove this line.

sebasLTripleTen reviewed (changes requested):
Hello, Michael!

Thank you for submitting your project. It's clear that a lot of work has been put into this project.

**Parts I really liked in your project:**

- Well-structured and clean code
- All variables, functions, and handlers are in the right places
- Great job handling potential errors at the end of your server requests.
- It's good that you're using `finally` blocks to reset the button text.

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

- All comments in the code, along with the ones below:
   - The code that checks the server response should be refactored to avoid repetition.
   - You need to close popups only in blocks then after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared, and the user could send the data again to the server), and if you change the button text you could see the change as well.
   - You shouldn't close the Edit Profile modal on unrelated requests.

There are also some comments for improvement marked as **"Could be improved"**. They don’t mean that you’ve done anything wrong, but they’re things we wanted to point out that can help you further develop your skills.

**Reminder:** For the project to be accepted, you must fix all **"Needs correcting"** comments. **"Could be improved"** comments are optional, so you can implement them if you choose.

It's fantastic to see how far you've come with this project. Even though there are areas for improvement, your hard work and dedication are evident. Remember that each step forward is progress. If you need help with any part of the feedback, our tutors are available to support you. Keep pushing forward and continue your impressive journey!

_If you have any questions or if any part of this review needs clarification, please don’t hesitate to reach out to me on Hub (Sebastian Lucero)._

malber71 committed:
fixed duplication in api.js

malber71 commented:
Removed the unnecessary close profile calls. This was the result of redundant copy paste. I added a submit button function to generically handle button submits and prevent redundancy in the .finally statements in the modal functions api calls.

malber71 committed:
revisions after first review

malber71 committed:
revisions after first review, one more revision

sebasLTripleTen commented on src/pages/index.js:
```diff
@@ -0,0 +1,433 @@
+
+  // card functionality
+
+  let selectedCard;
```
Currently, cards can't be deleted because the `selectedCard` and `selectedCardId` variables were moved inside the `getCardElement` function, making them inaccessible to the delete handler. 

To fix this, please move lines 329 and 330 out of `getCardElement` and define them in the global scope, so they can be accessed where needed—particularly in the delete functionality.

sebasLTripleTen reviewed (changes requested):
Hi!

Thank you for considering our recommendations. Your project has definitely improved. Most of the issues have been resolved, but there is something that still need attention:

- Currently, cards can't be deleted because the `selectedCard` and `selectedCardId` variables were moved inside the `getCardElement` function, making them inaccessible to the delete handler. 

Once this minor issue is addressed, your project will be ready for approval!

malber71 committed:
second review updates

malber71 committed:
updates since 2nd review

malber71 commented:
Moved the variables outside the function. I noticed that since we are working with the server the maxlength for card captions is 30 but in the original initial cards one of the card caption text is more than 30. I attempted to increase the maxlength to more than 30 but the server only accepts text length of 30 or less for this field. In the original design we were handling text more than 30 with text-overflow: ellipsis With the server limitation there is no way to test this on desktop view.

malber71 committed:
updates since 2nd review

malber71 committed:
updates since 2nd review

malber71 committed:
second review revisions

sebasLTripleTen commented:
> Moved the variables outside the function. I noticed that since we are working with the server the maxlength for card captions is 30 but in the original initial cards one of the card caption text is more than 30. I attempted to increase the maxlength to more than 30 but the server only accepts text length of 30 or less for this field. In the original design we were handling text more than 30 with text-overflow: ellipsis With the server limitation there is no way to test this on desktop view.

Hi! You can still test the `text-overflow: ellipsis` behavior by manually modifying the HTML in the browser using the DevTools panel.

For example, you can change the text content of a card directly in the DOM to a string longer than 30 characters and observe how the UI handles it.

From what I can see, it seems to be working as expected!

sebasLTripleTen reviewed (approved):
Great job, your project has been accepted!

Good luck with the next sprint, you’re doing awesome! 🚀
